core/src/main/scala/kafka/coordinator/group/CoordinatorLoaderImpl.scala

```diff
@@ -55,17 +56,21 @@ class CoordinatorLoaderImpl[T](
    * Loads the coordinator by reading all the records from the TopicPartition
    * and applying them to the Replayable object.
    *
-   * @param tp          The TopicPartition to read from.
-   * @param coordinator The object to apply records to.
+   * @param tp                      The TopicPartition to read from.
+   * @param coordinator             The object to apply records to.
+   * @param onLoadedBatch           Invoked when a batch was successfully loaded.
+   * @param onHighWatermarkUpdated  Invoked when the high watermark advanced.
    */
   override def load(
     tp: TopicPartition,
-    coordinator: CoordinatorPlayback[T]
+    coordinator: CoordinatorPlayback[T],
+    onLoadedBatch: Consumer[java.lang.Long],
+    onHighWatermarkUpdated: Consumer[java.lang.Long]
```

Comment:
This does not look good. Is there a reason why you did not use an interface which merges those three? In my opinion, it would be better to have an interface which defines:
* replay
* updateLastWrittenOffset
* updateLastCommittedOffset

We could re-purpose `CoordinatorPlayback` for this and introduce a new interface for the loader.

Reply:
i did not like the fact that we need to have a replay() method in both CoordinatorPlayback and CoordinatorShard. we cannot move replay() to just CoordinatorContext because it needs to access OffsetMetadataManager and GroupMetadataManager.

If you think that is fine, then I can make the changes. Is it not a good practice to use consumers like this?